ext/src/helpers/output_limited_buffer.rs

```diff
@@ -0,0 +1,50 @@
+use magnus::{value::InnerValue, value::Opaque, RString, Ruby};
+use std::io;
+
+pub struct OutputLimitedBuffer {
+    buffer: Opaque<RString>,
+    /// The maximum number of bytes that can be written to the output stream buffer.
+    capacity: usize,
+}
+
+impl OutputLimitedBuffer {
+    #[must_use]
+    pub fn new(buffer: Opaque<RString>, capacity: usize) -> Self {
+        Self { buffer, capacity }
+    }
+}
+
+impl io::Write for OutputLimitedBuffer {
+    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+        // Append a buffer to the string and truncate when hitting the capacity.
+        // We return the input buffer size regardless of whether we truncated or not to avoid a panic.
+        let ruby = Ruby::get().unwrap();
+
+        let mut inner_buffer = self.buffer.get_inner_with(&ruby);
+
+        if buf.is_empty() {
+            return Ok(buf.len());
+        }
+
+        if inner_buffer.len() + buf.len() > self.capacity {
```

Comment:
Both of these `.len()`s return `usize`, I wonder if we should instead do a [`checked_add` ](https://doc.rust-lang.org/std/primitive.usize.html#method.checked_add)and handle the overflow case.